lib/stream.js

```diff
@@ -1,17 +1,33 @@
 import {createReadStream, readFileSync} from 'node:fs';
+import {Readable} from 'node:stream';
 import {setTimeout} from 'node:timers/promises';
 import {isStream} from 'is-stream';
 import getStream, {getStreamAsBuffer} from 'get-stream';
 import mergeStreams from '@sindresorhus/merge-streams';
+import {stdioHasIterableStdin} from './stdio.js';
 
-export const validateInputOptions = ({input, inputFile}) => {
+export const validateInputOptions = ({input, inputFile}, stdio) => {
 	if (input !== undefined && inputFile !== undefined) {
 		throw new TypeError('The `input` and `inputFile` options cannot be both set.');
 	}
+
+	if (stdioHasIterableStdin(stdio)) {
+		if (input !== undefined) {
+			throw new TypeError('The `stdin` option cannot be an iterable when the `input` option is set.');
+		}
+
+		if (inputFile !== undefined) {
+			throw new TypeError('The `stdin` option cannot be an iterable when the `inputFile` option is set.');
+		}
+	}
 };
 
 // `input` and `inputFile` option in sync mode
-export const handleInputSync = ({input, inputFile}) => {
+export const handleInputSync = ({input, inputFile}, stdio) => {
+	if (stdioHasIterableStdin(stdio)) {
+		throw new TypeError('The `stdin` option cannot be an iterable in sync mode');
```

Comment:
Due to [the above](https://github.com/sindresorhus/execa/pull/604/files#r1427496220), this PR's logic requires stream piping, which cannot work in sync mode.

Reply:
This should be noted in the docs.

Reply:
Fixed. :+1: